codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -317,125 +314,60 @@ private Http2Stream requireStream(int streamId) {
-                    }
-                    int writeablePadding = Math.min(allowedFrameSize - writeableData, padding);
-                    padding -= writeablePadding;
-                    bytesWritten += writeableData + writeablePadding;
-                    ChannelPromise writePromise;
-                    if (size == bytesWritten && !promise.isVoid()) {
-                        // Can use the original promise if it's the last write
-                        writePromise = promise;
-                    } else {
-                        // Create a new promise and listen to it for failure
-                        writePromise = ctx.newPromise();
-                        writePromise.addListener(this);
-                    }
-                    if (toWrite instanceof SlicedByteBuf && data instanceof CompositeByteBuf) {
-                        // If we're writing a subset of a composite buffer then we want to release
-                        // any underlying buffers that have been consumed. CompositeByteBuf only releases
-                        // underlying buffers on write if all of its data has been consumed and its refCnt becomes
-                        // 0.
-                        final CompositeByteBuf toFree = (CompositeByteBuf) data;
-                        writePromise.addListener(new ChannelFutureListener() {
-                            @Override
-                            public void operationComplete(ChannelFuture future) throws Exception {
-                                toFree.discardReadComponents();
-                            }
-                        });
-                    }
-                    frameWriter().writeData(ctx, stream.id(), toWrite, writeablePadding,
-                            size == bytesWritten && endOfStream, writePromise);
-                } while (size != bytesWritten && allowedBytes > bytesWritten);
-            } finally {
-                size -= bytesWritten;
-            }
+            int maxFrameSize = frameWriter().configuration().frameSizePolicy().maxFrameSize();
+            int bytesWritten = 0;
+            do {
+                int allowedFrameSize = Math.min(maxFrameSize, allowedBytes - bytesWritten);
+                int writeableData = Math.min(queue.readableBytes(), allowedFrameSize);
+                ChannelPromise writePromise = ctx.newPromise();
+                writePromise.addListener(this);
+                ByteBuf toWrite = queue.remove(writeableData, writePromise);
+
+                int writeablePadding = Math.min(allowedFrameSize - writeableData, padding);
+                padding -= writeablePadding;
+                bytesWritten += writeableData + writeablePadding;
+                frameWriter().writeData(ctx, stream.id(), toWrite, writeablePadding,
+                        queue.readableBytes() + padding == 0 && endOfStream, writePromise);
+            } while (queue.readableBytes() + padding > 0 && allowedBytes > bytesWritten);
```

Comment:
I think this is covered but just to make sure ... if there are zero-length data frames in the queue, will we still pull them out if `allowedBytes == 0`? I think since this is a `do...while` loop it should work.

Hmm ... I wonder if the check at the top of the method is an issue in this regard. If we're told to write zero bytes, we won't enter the loop.